### PR TITLE
Feature/51 No more NiFi HTTP edition

### PR DESCRIPTION
**Pull request by jimmarino**
Transfer functions are a simple way to create data flow extensions for EDC: implement an HTTP endpoint that handles data transfer according to the documentation and configure the Transfer Functions extension.

The HTTP endpoint can be written in any language and hosted on cloud infrastructure (such as an Azure Function or AWS Lambda) for scale-out. In the future, we can support [Keda](https://keda.sh/) for on-premise or DIY scale-out.

### CODE REVIEW

File: extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionConfiguration.java [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+import java.util.function.Supplier;
+
+/**
+ * Configuration for the HTTP data flow controller and status checker.
+ */
+public class HttpFunctionConfiguration {
+    private String transferEndpoint;
+    private String checkEndpoint;
+    private Set<String> protocols;
+    private Supplier<OkHttpClient> clientSupplier;
```
Comment (beardyinc) [inline extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionConfiguration.java:30]: Why not directly use the `OkHttpClient`? It is resolvable via the `ServiceExtensionContext`

File: extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreServiceExtension.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+        var httpClient = createHttpClient(context);
+        context.registerService(TransferFunctionInterceptorRegistry.class, httpClient::addInterceptor);
+
+        var typeManager = context.getTypeManager();
+        var transferEndpoint = context.getSetting(TRANSFER_URL_KEY, DEFAULT_LOCAL_TRANSFER_URL);
+        var checkEndpoint = context.getSetting(CHECK_URL_KEY, DEFAULT_LOCAL_CHECK_URL);
+        var configuration = HttpFunctionConfiguration.Builder.newInstance()
+                .transferEndpoint(transferEndpoint)
+                .checkEndpoint(checkEndpoint)
+                .clientSupplier(httpClient::build)
```
Comment (beardyinc) [inline extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreServiceExtension.java:136]: the `httpClient` can be directly resolved using the context. In that case a `requires` dependecy to `"dataspaceconnector:http-client"` must be defined.

```suggestion
                .clientSupplier(context.getService(OkHttpClient.class)
```

Using one HttpClient globally has the upside of central configuration (timeouts, etc.) and error handling.

File: extensions/transfer-functions/transfer-functions-core/src/test/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreHttpTest.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+                .destinationType("foo")
+                .dataEntry(dataEntry)
+                .managedResources(false)
+                .dataDestination(DataAddress.Builder.newInstance().type("test-protocol1").build())
+                .connectorId("test").build();
+
+        processManager.initiateProviderRequest(dataRequest);
+
+        latch.await(1, TimeUnit.MINUTES);
+        Assertions.assertTrue(invoked.get());
```
Comment (beardyinc) [inline extensions/transfer-functions/transfer-functions-core/src/test/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreHttpTest.java:77]: we can always `assertTrue(latch.await(...))`, which would save us the `assertTrue(invoked.get())`. 
```suggestion
       assertThat(latch.await(1, TimeUnit.MINUTES)).isTrue()
```
Also, I think we should try and stick with one assertion framework (here: AssertJ)

File: extensions/transfer-functions/transfer-functions-core/src/test/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreLocalTest.java [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+                .destinationType("foo")
+                .dataEntry(dataEntry)
+                .managedResources(false)
+                .dataDestination(DataAddress.Builder.newInstance().type("test-protocol1").build())
+                .connectorId("test").build();
+
+        processManager.initiateProviderRequest(dataRequest);
+
+        latch.await(1, TimeUnit.MINUTES);
+        Assertions.assertTrue(invoked);
```
Comment (beardyinc) [inline extensions/transfer-functions/transfer-functions-core/src/test/java/org/eclipse/dataspaceconnector/transfer/functions/core/TransferFunctionsCoreLocalTest.java:62]: we can always assertTrue(latch.await(...)), which would save us the assertTrue(invoked.get()).
```suggestion
        assertThat(latch.await(1, TimeUnit.MINUTES)).isTrue();
```
Also, I think we should try and stick with one assertion framework (here: AssertJ)

File: launchers/junit/src/testFixtures/java/org/eclipse/dataspaceconnector/junit/launcher/MockVault.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+import org.eclipse.dataspaceconnector.spi.security.VaultResponse;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * An in-memory vault for testing.
+ */
+public class MockVault implements Vault {
```
Comment (beardyinc) [inline launchers/junit/src/testFixtures/java/org/eclipse/dataspaceconnector/junit/launcher/MockVault.java:26]: This class seems not to be used anywhere. There is another `MockVault` in `DemoProtocolsTransferExtensionTest.java`, which looks pretty much the same.

**Review by beardyinc — CHANGES_REQUESTED** [review]

File: launchers/junit/src/testFixtures/java/org/eclipse/dataspaceconnector/junit/launcher/MockVault.java (hunk 5 above)
Comment (jimmarino) [reply launchers/junit/src/testFixtures/java/org/eclipse/dataspaceconnector/junit/launcher/MockVault.java:26]: Oh this looks like a rebase problem. I'll delete it

File: extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionConfiguration.java (hunk 1 above)
Comment (jimmarino) [reply extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionConfiguration.java:30]: I need to document this...The extension will allow the use of custom interceptors to set headers for things like authorization.

**Comment by beardyinc** [comment]
good to go

File: extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionDataFlowController.java [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    }
+
+    @Override
+    public boolean canHandle(DataRequest dataRequest) {
+        return protocols.contains(dataRequest.getDestinationType());
+    }
+
+    @Override
+    public @NotNull DataFlowInitiateResponse initiateFlow(DataRequest dataRequest) {
+        var requestBody = RequestBody.create(typeManager.writeValueAsString(dataRequest), JSON);
```
Comment (MoritzKeppler) [inline extensions/transfer-functions/transfer-functions-core/src/main/java/org/eclipse/dataspaceconnector/transfer/functions/core/flow/http/HttpFunctionDataFlowController.java:64]: I always get a little stomach ache when passing internal objects to processes outside without beeing in control what's actually inside - do we know that a DataRequest contains solely information the other party is allowed to see?
Would it make sense to document that usage in the DataRequest class? e.g. "just store information that can be passed to ..."

**Review by MoritzKeppler — APPROVED** [review]

**Pull request merged**